src/lib.rs

```diff
@@ -6,18 +6,203 @@ An experimental x86_64 bootloader that works on both BIOS and UEFI systems.
 
 #[cfg(feature = "bios")]
 mod bios;
-mod fat;
+#[cfg(feature = "uefi")]
+mod gpt;
+#[cfg(feature = "bios")]
+mod mbr;
 #[cfg(feature = "uefi")]
 mod uefi;
 
+#[cfg(feature = "uefi")]
+pub use uefi::UefiBoot;
+
 #[cfg(feature = "bios")]
 pub use bios::BiosBoot;
 
-#[cfg(feature = "uefi")]
-pub use uefi::UefiBoot;
+mod fat;
+
+use std::{
+    collections::BTreeMap,
+    env::temp_dir,
+    fs,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use anyhow::Context;
+
+use tempfile::NamedTempFile;
 
 pub use bootloader_boot_config::BootConfig;
 
 const KERNEL_FILE_NAME: &str = "kernel-x86_64";
 const RAMDISK_FILE_NAME: &str = "ramdisk";
 const CONFIG_FILE_NAME: &str = "boot.json";
+
+#[derive(Clone)]
+struct DiskImageFile {
+    source: PathBuf,
+    destination: String,
+}
+
+/// DiskImageBuilder helps create disk images for a specified set of files.
+/// It can currently create MBR (BIOS), GPT (UEFI), and TFTP (UEFI) images.
+pub struct DiskImageBuilder {
+    files: Vec<DiskImageFile>,
+}
+
+impl DiskImageBuilder {
+    /// Create a new instance of DiskImageBuilder, with the specified kernel.
+    pub fn new(kernel: &Path) -> Self {
+        let mut obj = Self::empty();
+        obj.set_kernel(kernel);
+        obj
+    }
+
+    /// Create a new, empty instance of DiskImageBuilder
+    pub fn empty() -> Self {
+        Self { files: Vec::new() }
+    }
+
+    /// Add or replace a kernel to be included in the final image.
+    pub fn set_kernel(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, KERNEL_FILE_NAME)
+    }
+
+    /// Add or replace a ramdisk to be included in the final image.
+    pub fn set_ramdisk(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, RAMDISK_FILE_NAME)
+    }
+
+    /// Configures the runtime behavior of the bootloader.
+    pub fn set_boot_config(&mut self, boot_config: &BootConfig) -> &mut Self {
+        let json =
+            serde_json::to_string_pretty(boot_config).expect("failed to serialize BootConfig");
+        let bytes = json.as_bytes();
+        self.add_or_replace_file_byte_content(bytes, CONFIG_FILE_NAME)
+    }
+
+    /// Add or replace a file from a byte array
+    pub fn add_or_replace_file_byte_content(&mut self, data: &[u8], target: &str) -> &mut Self {
```

Comment:
The name seems a bit long. How about `set_file_with_content`? Also, I would prefer to swap the argument order (target path first, data second).

Reply:
Sounds good to me!